source/game.production.script.bmx

```diff
@@ -155,6 +155,9 @@ Type TScriptCollection Extends TGameObjectCollection
 				filter.skipNotAvailable = True
 				template = GetScriptTemplateCollection().GetRandomByFilter(filter)
 			EndIf
+rem
```

Comment:
Generell gibt es nichts gegen das Reduzieren von Code-Dopplung und die Vereinfachung der Vermeidung von Titelduplikaten einzuwenden.

Ich glaube aber, dass das Refactoring ins seiner aktuellen Form nach hinten losgehen könnte. Wenn sich ausschließlich das Template selbst um die Titeleindeutigkeit kümmert (und Counter anhängt), kann es sein dass die getRandom-Methode schlecht Scripte auswählt.
Bsp. aus unerfindlichen Gründen wird beim ersten getRandomByFilter immer dasselbe Template ausgewählt. Dann würden nur noch Titel mit Counter erzeugt. Es geht ja nicht nur darum, dass die Titel möglichst eindeutig sind sondern auch darum, dass unterschiedliche Vorlagen zum Zug kommen.

Vorschlag: mindestens hier verlassen wir uns nicht nur auf den vom Template eindeutig generierten Titel, sondern prüfen den Titel auf Counter (endet mit # und Zahl) und setzen in dem Fall das Template auf null für einen neuen Auswahlversuch.

Reply:
? Die Titelduplikate vermeidet "TScript", nicht "TScriptTemplate" (dem Template ist das egal ...das liefert ja die Rohdaten bzw evaluierte Variablen)

Reply:
Durch das Reinziehen des Titelschutzes hat der Aufrufer nicht mehr die Information, dass ein Counter angehängt wird - vielleicht hätte er es ja gar nicht gewollt und stattdessen ein anderes Template verwendet.
Auf alle Fälle nicht wie aktuell im Code beschrieben den auskommentierten Teil rauswerfen, sondern ein dickes TODO ran.
Ich könnte mir ein noSave-Feld am Script vorstellen, das für den Aufrufer ein Indikator ist, dass der Titel um einen Counter ergänzt wurde. Dann muss man nicht den Titel prüfen.

Reply:
Du beziehst Dich also direkt auf den "ge-rem-ten" Block mit dem Protected-Check fuer Episoden?

Das Problem ist aber ja, dass der Titel bzw die Beschreibung abhaengig von anderen Drehbuechern "der Serie" sein kann.
Sprich die Templates koennten nur "fuer sich alleinstehende" Titel/Beschreibungen generieren.

Also Bspweise `Best of ${.self:"rolescount"}`, `${.self:blocks} Stunden Party Non Stop!` oder so. Fuer das andere muessten die Templates ueber "Kinder-Templates" Informationen ziehen usw. ...hm.


Momentan muesste dieser Check nur dort drin stattfinden:
```
			If template
				Local script:TScript = TScript.CreateFromTemplate(template,True)

...hier
If template.GetProductionLimitMax() > 1
...
	If Not tempName Then template = Null
EndIf


				script.SetOwner(TOwnedGameObject.OWNER_NOBODY)
				Add(script)
				Return script
			EndIf
```

Ich verstehe aber, dass natuerlich die Skripterstellung schon "mehr macht", als das alte Template.GenerateFinalTitle().
Aber wir koennen gern `_GenerateTitleFromTemplate()` aufbohren, so dass nicht nur ein "true/false" zurueckgegeben wird, sondern man auch steuern kann, ob "#2, #3..." erlaubt werden soll oder nicht - und was genau feschlug.
Dann wuerde man obiges kombinieren koennen...

```
	Method GenerateRandom:TScript(avoidTemplateIDs:Int[])
		Local template:TScriptTemplate
		Local filter:SScriptTemplateFilter = new SScriptTemplateFilter
		filter.biggerStudioAllowedChance = 10 + 5 * (GetWorldTime().GetDaysRun())
		'iterate several times in case template with production limit and protected title is involved
		For Local i:Int = 0 Until 20
				Local script:TScript = TScript.CreateFromTemplate(template,True)

				'If the template has a production limit, ensure no protected title is used
				If template.GetProductionLimitMax() > 1
					Local titleOK:Int = False
					For Local i:Int = 0 Until 20
						template.reset()
						script._GenerateTitleFromTemplate()
						If Not GetScriptCollection().IsTitleProtected(script.title)
							titleOK = True
							Exit
						EndIf
					Next
					If Not titleOK Then script = Null
				EndIf

				If script
					script.SetOwner(TOwnedGameObject.OWNER_NOBODY)
					Add(script)
				EndIf
				Return script
...
```